### PR TITLE
HPCC-26626 Add support for hardware cost section in configmgr

### PR DESCRIPTION
**Pull request by kenrowland**
Added support for the cost element in the hardware section

Signed-off-by: Ken Rowland kennethrowland@lexisnexisrisk.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: initfiles/componentfiles/configxml/environment.xsd.in [hunk 1]
```diff
@@ -121,6 +121,17 @@
                                     <xs:attribute name="domain" type="xs:string"/>
                                     <xs:attribute name="computerType" type="xs:string"/>
                                 </xs:complexType>
+
+                                <xs:element name="cost">
```
Comment (rpastrana) [inline initfiles/componentfiles/configxml/environment.xsd.in:125]: It seems other elements are capitalized

File: initfiles/componentfiles/configxml/environment.xsd.in [hunk 2]
```diff
@@ -121,6 +121,17 @@
...
+                                    <xs:complexType>
+                                        <xs:attribute name="memoryLocal" type="xs:string"/>
```
Comment (rpastrana) [inline initfiles/componentfiles/configxml/environment.xsd.in:127]: Perhaps in context it becomes obvious, but at least for me these attribute names are not entirely self-documenting.

File: deployment/deployutils/deployutils.cpp [hunk 3]
```diff
@@ -2066,6 +2090,13 @@ bool generateHardwareHeaders(const IPropertyTree* pEnv, StringBuffer& sbDefn, bo
     jsStrBuf.appendf("colIndex['directoryNAS']=%d;", index++);
     jsStrBuf.appendf("colIndex['traceNAS']=%d;", index++);
 
+    index=0;
+    jsStrBuf.appendf("colIndex['memoryLocalecost']=%d;", index++);
```
Comment (rpastrana) [inline deployment/deployutils/deployutils.cpp:2094]: It seems "cost" should be capitalized in all of these calls.

File: deployment/deployutils/deployutils.cpp [hunk 4]
```diff
@@ -2031,16 +2046,25 @@ bool generateHardwareHeaders(const IPropertyTree* pEnv, StringBuffer& sbDefn, bo
...
     jsStrBuf.append("compTabs['Hardware'][compTabs['Hardware'].length]= 'Domains';");
     jsStrBuf.append("compTabs['Hardware'][compTabs['Hardware'].length]= 'Computers';");
     jsStrBuf.append("compTabs['Hardware'][compTabs['Hardware'].length]= 'NAS';");
+    jsStrBuf.append("compTabs['Hardware'][compTabs['Hardware'].length]= 'cost';");
     jsStrBuf.append("compTabToNode['Computer Types']= 'ComputerType';");
     jsStrBuf.append("compTabToNode['Switches']= 'Switch';");
     jsStrBuf.append("compTabToNode['Domains']= 'Domain';");
     jsStrBuf.append("compTabToNode['Computers']= 'Computer';");
     jsStrBuf.append("compTabToNode['NAS']= 'NAS';");
+    jsStrBuf.append("compTabToNode['cost']= 'cost';");
```
Comment (rpastrana) [inline deployment/deployutils/deployutils.cpp:2067]: Cost might need to be capitalized based on local trends

File: deployment/deployutils/deployutils.cpp [hunk 5]
```diff
@@ -2004,6 +2004,21 @@ bool generateHardwareHeaders(const IPropertyTree* pEnv, StringBuffer& sbDefn, bo
       xpath.clear().append(XML_ATTR_TRACE);
       pNAS->addProp(xpath, sbdefaultValue.str());
     }
+
+    //<cost moneyLocale="en_US.UTF-8" perCpu="0.113" storageAtRest="0.0135" storageReads="0.0485" storageWrites="0.0038"/>
+
+    IPropertyTree* pCost = pCompTree->addPropTree("cost", createPTree());
+    xpath.clear().append("@memoryLocale");
```
Comment (rpastrana) [inline deployment/deployutils/deployutils.cpp:2011]: is this attribute supposed to be money locale rather than  memory locale?

**Review by rpastrana — CHANGES_REQUESTED** [review]
@kenrowland a few comments, mostly trivial

File: deployment/deployutils/deployutils.cpp (hunk 5 above)
Comment (kenrowland) [reply deployment/deployutils/deployutils.cpp:2011]: Shamser added it as a lower case. See https://github.com/hpcc-systems/HPCC-Platform/pull/15412/files.

I just duplicated it here.

**Review by kenrowland — COMMENTED** [review]
@rpastrana Please see comment.

File: deployment/deployutils/deployutils.cpp (hunk 5 above)
Comment (rpastrana) [reply deployment/deployutils/deployutils.cpp:2011]: @kenrowland that seems unfortunate but not critical. At this point it's a matter of following the standards inherited in these files vs effort to change in the code. I'll leave it up to you and @shamser

**Comment by kenrowland** [comment]
@rpastrana Can you approve based on deferring to lowercase yaml convention?

**Comment by rpastrana** [comment]
> 
> 
> @rpastrana Can you approve based on deferring to lowercase yaml convention?

I didn't see a reply to the "memoryLocale" question.

File: deployment/deployutils/deployutils.cpp (hunk 4 above)
Comment (kenrowland) [reply deployment/deployutils/deployutils.cpp:2067]: Follows yaml conventions.

File: initfiles/componentfiles/configxml/environment.xsd.in (hunk 1 above)
Comment (kenrowland) [reply initfiles/componentfiles/configxml/environment.xsd.in:125]: Follows yaml conventions.

**Review by rpastrana — APPROVED** [review]

**Review by shamser — APPROVED** [review]

**Pull request merged**